system-parachains/collectives/collectives-polkadot/src/impls.rs

```diff
@@ -159,6 +164,53 @@ impl PrivilegeCmp<OriginCaller> for EqualOrGreatestRootCmp {
 	}
 }
 
+/// Contains a system-level sibling parachain.
+pub struct IsSiblingSystemParachain<ParaId, SelfParaId>(PhantomData<(ParaId, SelfParaId)>);
+impl<ParaId: IsSystem + From<u32> + Eq, SelfParaId: Get<ParaId>> Contains<Location>
+	for IsSiblingSystemParachain<ParaId, SelfParaId>
+{
+	fn contains(l: &Location) -> bool {
+		matches!(
+			l.unpack(),
+			(1,	[Parachain(id)]) if ParaId::from(*id).is_system() && SelfParaId::get() != ParaId::from(*id),
+		)
+	}
+}
+
+/// Determines if the given `asset_kind` is a native asset. If it is, returns the balance without
+/// conversion; otherwise, delegates to the implementation specified by `I`.
+pub struct NativeOnSiblingParachain<I, SelfParaId>(PhantomData<(I, SelfParaId)>);
+impl<I, SelfParaId> ConversionFromAssetBalance<Balance, VersionedLocatableAsset, Balance>
+	for NativeOnSiblingParachain<I, SelfParaId>
+where
+	I: ConversionFromAssetBalance<Balance, VersionedLocatableAsset, Balance>,
+	SelfParaId: Get<ParaId>,
+{
+	type Error = ();
+	fn from_asset_balance(
+		balance: Balance,
+		asset_kind: VersionedLocatableAsset,
+	) -> Result<Balance, Self::Error> {
+		use VersionedLocatableAsset::*;
+		let (location, asset_id) = match asset_kind.clone() {
+			V3 { location, asset_id } => (location.try_into()?, asset_id.try_into()?),
+			V4 { location, asset_id } => (location, asset_id),
+		};
+
+		if asset_id.0.contains_parents_only(1) &&
+			IsSiblingSystemParachain::<ParaId, SelfParaId>::contains(&location)
```

Comment:
I think that the `NativeOnSiblingParachain` for collectives and `NativeOnSystemParachain` for relays are "almost" the same implementation and differ only with the different `::contains(&location)` implementation for matching system parachains.

I'm also not sure about the same `asset_id.0.contains_parents_only(1)` in both implementations, but regardless of that, they are exactly the same. I would suggest creating just one generic implementation and placing it in the `polkadot-sdk` repository somewhere. As a temporary solution, with a TODO, it could be placed in the `system-parachains-constants` repository. Yes, it may be ugly, but it serves the purpose. Alternatively, you could copy the exact same generic implementation where needed and remove it later.

Suggested generic implementation:

```
pub struct NativeOnSystemParachainElseConvertWith<I, N, IsSystemParachain>(PhantomData<(I, N, IsSystemParachain)>);
impl<I, N, IsSystemParachain> ConversionFromAssetBalance<Balance, VersionedLocatableAsset, Balance>
	for NativeOnSystemParachainElseConvertWith<I, N, IsSystemParachain>
where
	I: ConversionFromAssetBalance<Balance, VersionedLocatableAsset, Balance>,
        IsNativeAsset: Contains<Location> // <-- or maybe Contains<AssetId>?, we have already matchers for Contains<Location>
	IsSystemParachain: Contains<Location>,
{
	type Error = ();
	fn from_asset_balance(
		balance: Balance,
		asset_kind: VersionedLocatableAsset,
	) -> Result<Balance, Self::Error> {
		use VersionedLocatableAsset::*;
		let (location, asset_id) = match asset_kind.clone() {
			V3 { location, asset_id } => (location.try_into()?, asset_id.try_into()?),
			V4 { location, asset_id } => (location, asset_id),
		};

		if IsNativeAsset::contains(asset_id.0) && IsSystemParachain::contains(&location)
                {
			Ok(balance)
		} else {
			I::from_asset_balance(balance, asset_kind).map_err(|_| ())
		}
	}
	#[cfg(feature = "runtime-benchmarks")]
	fn ensure_successful(asset_kind: VersionedLocatableAsset) {
		I::ensure_successful(asset_kind)
	}
}                
```

so with this one generic implementation, it can be used for both:
```
// for collectives
type BalanceConverter = NativeOnSystemParachainElseConvertWith<
     AssetRate,
     Equals<xcm_config::DotLocation>
     IsSiblingSystemParachain<ParaId, ParachainInfo>,
>;


// for relays
type BalanceConverter = NativeOnSystemParachainElseConvertWith<
     AssetRate,
     ( 
         Equals<DotLocation>
         Equals<DotLocationWithParent1>   // <- kind of OR matching
     ),
     IsChildSystemParachain::<ParaId>,
>;

```

Reply:
PR description referes to polkadot-sdk version - https://github.com/paritytech/polkadot-sdk/pull/3659